parquet/src/main/java/org/apache/iceberg/parquet/ParquetTypeWithPartnerVisitor.java

```diff
@@ -0,0 +1,250 @@
+        "Invalid map: top-level group is repeated: %s", group);
+    Preconditions.checkArgument(group.getFieldCount() == 1,
+        "Invalid map: does not contain single repeated field: %s", group);
+
+    GroupType repeatedKeyValue = group.getType(0).asGroupType();
+    Preconditions.checkArgument(repeatedKeyValue.isRepetition(Type.Repetition.REPEATED),
+        "Invalid map: inner group is not repeated");
+    Preconditions.checkArgument(repeatedKeyValue.getFieldCount() <= 2,
+        "Invalid map: repeated group does not have 2 fields");
+
+    visitor.beforeRepeatedKeyValue(repeatedKeyValue);
+    try {
+      T keyResult = null;
+      T valueResult = null;
+      switch (repeatedKeyValue.getFieldCount()) {
+        case 2:
+          // if there are 2 fields, both key and value are projected
+          Type keyType = repeatedKeyValue.getType(0);
+          visitor.beforeKeyField(keyType);
+          try {
+            keyResult = visit(visitor.mapKeyType(map), keyType, visitor);
+          } finally {
+            visitor.afterKeyField(keyType);
+          }
+          Type valueType = repeatedKeyValue.getType(1);
+          visitor.beforeValueField(valueType);
+          try {
+            valueResult = visit(visitor.mapValueType(map), repeatedKeyValue.getType(1), visitor);
+          } finally {
+            visitor.afterValueField(valueType);
+          }
+          break;
+        case 1:
+          // if there is just one, use the name to determine what it is
+          Type keyOrValue = repeatedKeyValue.getType(0);
+          if (keyOrValue.getName().equalsIgnoreCase("key")) {
+            visitor.beforeKeyField(keyOrValue);
+            try {
+              keyResult = visit(visitor.mapKeyType(map), keyOrValue, visitor);
+            } finally {
+              visitor.afterKeyField(keyOrValue);
+            }
+            // value result remains null
+          } else {
+            visitor.beforeValueField(keyOrValue);
+            try {
+              valueResult = visit(visitor.mapValueType(map), keyOrValue, visitor);
+            } finally {
+              visitor.afterValueField(keyOrValue);
+            }
+            // key result remains null
+          }
+          break;
+        default:
+          // both results will remain null
+      }
+
+      return visitor.map(map, group, keyResult, valueResult);
+
+    } finally {
+      visitor.afterRepeatedKeyValue(repeatedKeyValue);
+    }
+  }
+
+  protected static <P, T> List<T> visitFields(P struct, GroupType group, ParquetTypeWithPartnerVisitor<P, T> visitor) {
+    List<T> results = Lists.newArrayListWithExpectedSize(group.getFieldCount());
+    for (int i = 0; i < group.getFieldCount(); i += 1) {
+      Type field = group.getFields().get(i);
+      visitor.beforeField(field);
+      Integer fieldId = field.getId() == null ? null : field.getId().intValue();
+      results.add(visit(visitor.fieldType(struct, i, fieldId), field, visitor));
+      visitor.afterField(field);
```

Comment:
Q:  should we use a `try-finally` block here for the `beforeField & afterField` ?

Reply:
Yes, I think so. Nice catch!